### PR TITLE
PrepareData + DataOverTime Entity

### LINKED ISSUE #34: PrepareDataService interface

### PR DESCRIPTION
**Pull request by dottore1**
Closes #34

## Description
Implements PrepareData abstract class, to prepare DataOverTime to produce a batch report.
Implements the DataOverTime entity used in PrepareData.

## Checklist:
- [X] My code follows the code style of this project.
- [x] I have added tests to cover my changes, if necessary.
- [x] All tests pass (existing and potential new).
- [x] Sonar check has passed
- [x] No codesmells or bugs

## Other Relevant Information
Provide any other important details below.

### CODE REVIEW

**Review by Wylardt — APPROVED** [review]

**Comment by Lawst96** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL) [2 Code Smells](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_duplicated_lines_density&view=list)

Oh, how the [turntables](https://github.com/Semesterproject4/back-end/pull/62#issuecomment-834300629)

**Comment by Kechr** [comment]
> > SonarCloud Quality Gate failed.
> > [Bug](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=BUG)
> > [Vulnerability](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=VULNERABILITY)
> > [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=SECURITY_HOTSPOT)
> > [Code Smell](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL) [2 Code Smells](https://sonarcloud.io/project/issues?id=Semesterproject4_back-end&pullRequest=64&resolved=false&types=CODE_SMELL)
> > [](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_coverage&view=list)
> > [](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Semesterproject4_back-end&pullRequest=64&metric=new_duplicated_lines_density&view=list)
> 
> Oh, how the [turntables](https://github.com/Semesterproject4/back-end/pull/62#issuecomment-834300629)

Hvis du kigger på hvilke code smells det er, ville du ikke komme med sådan en kommentar. Den første er at vi skal lave vores abstrakte klasse til et interface, da den ikke bliver brugt endnu. Den anden er at vi skal bruge 'var' frem for den reelle objekttype, hvilket vi ikke mener er hensigtsmæssigt, når vi koder i Java. Den sidste er jeg ved at finde en løsning på, så vi ekskluderer den regel fra sonar.

File: BrewMesModulesParent/batch/src/main/java/dk/sdu/mmmi/batch/PrepareData.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import com.brewmes.common.util.Products;
+
+import java.time.LocalDateTime;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public abstract class PrepareData implements IBatchReportService {
+
+    private double calculateOee(int amountToProduce, double machineSpeed,  int maxMachineSpeed, int acceptedProducts) {
```
Comment (Teyson) [inline BrewMesModulesParent/batch/src/main/java/dk/sdu/mmmi/batch/PrepareData.java:15]: Tilføj JavaDoc på de metoder der ikke er fra IBatchReportService

File: BrewMesModulesParent/batch/src/main/java/dk/sdu/mmmi/batch/PrepareData.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import com.brewmes.common.entities.MachineData;
+import com.brewmes.common.services.IBatchReportService;
+import com.brewmes.common.util.Products;
+
+import java.time.LocalDateTime;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public abstract class PrepareData implements IBatchReportService {
```
Comment (Teyson) [inline BrewMesModulesParent/batch/src/main/java/dk/sdu/mmmi/batch/PrepareData.java:13]: Hvordan implement'er i et interface uden at implementere dens metode??

File: BrewMesModulesParent/batch/src/test/java/dk/sdu/mmmi/batch/PrepareDataTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        };
+
+        overTime = prepareData.prepareData(batch);
+    }
+
+    @Test
+    void calculateOeeTest() {
+        double plannedProductionTime = (AMOUNT_TO_PRODUCE / DESIRED_SPEED) * 60.0;
+        double idealCycleTime = (1.0 / MAX_MACHINE_SPEED) * 60;
+        double expected = ((ACCEPTED_PRODUCTS * idealCycleTime) / plannedProductionTime) * 100;
```
Comment (Teyson) [inline BrewMesModulesParent/batch/src/test/java/dk/sdu/mmmi/batch/PrepareDataTest.java:54]: Den går sgu ikke drenge xD I kan ikke bare copy+paste udregningen fra koden ind og bruge den som test. Det fanger jo ikke eventuelle fejl

**Review by Teyson — APPROVED** [review]

**Review by Teyson — CHANGES_REQUESTED** [review]
Se ovenstående kommentarer

**Review by Lawst96 — APPROVED** [review]

**Review by Teyson — APPROVED** [review]
Se ovenstående kommentarer

**Review by Kaszz — APPROVED** [review]

**Pull request merged**